### PR TITLE
Odo HashValidation implemented

### PR DESCRIPTION
**Pull request by SmartArray**
- Invalid datatype for Permutation function fixed
- Added test cases
- Fixed the last odo bugs
- Optimized some code

### CODE REVIEW

File: crypto/odocrypt.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+typedef struct {
+    uint64_t current;
+    uint64_t multiplicand;
+    uint64_t addend;
+} OdoRandom;
+
+const static uint64_t BASE_MULTIPLICAND = 6364136223846793005ull;
+const static uint64_t BASE_ADDEND = 1442695040888963407ull;
+
+int Odocrypt_Hash(OdoStruct* odo, const char* pbegin, const char* pend, char* output);
```
Comment (MentalCollatz) [inline crypto/odocrypt.h:63]: Capitalization of the "c" is inconsistent here.

File: crypto/odocrypt.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#ifndef HASH_ODO
+#define HASH_ODO
+
+#include <stdlib.h>
+
+#define ODOCRYPT_CHAPECHANGE_INTERVAL (10*24*60*60)
+#define ODOCRYPT_DIGEST_SIZE          (80)
+
+// Number of rounds.
+const static int ROUNDS = 84;
```
Comment (MentalCollatz) [inline crypto/odocrypt.h]: It would be slightly better to add a prefix to all of the constants.

File: crypto/odocrypt.c [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+    for (int i = 0; i < STATE_SIZE; i++)
+    {
+        for (int j = 0; j < 8; j++)
+        {
+            bytes[8*i + j] = (state[i] >> (8*j)) & 0xff;
+        }
+    }
+}
+
+void OdoCrypt_ApplySboxes(uint64_t* state, size_t sz1, size_t sz2, const uint8_t sbox1[][sz1], const uint16_t sbox2[][sz2])
```
Comment (MentalCollatz) [inline crypto/odocrypt.c]: Are params 2 and 3 necessary?  They will always be `(1 << SMALL_SBOX_WIDTH)` and `(1 << LARGE_SBOX_WIDTH)`, respectively.

File: BRMerkleBlock.c [hunk 4]
```diff
@@ -139,25 +139,38 @@ BRMerkleBlock *BRMerkleBlockParse(const uint8_t *buf, size_t bufLen)
...
 
-            default:
+            case BLOCK_VERSION_SCRYPT:
+                // void BRScrypt(void *dk, size_t dkLen, const void *pw, size_t pwLen, const void *salt, size_t saltLen, unsigned n, unsigned r, unsigned p)
                 BRScrypt(&block->powHash, sizeof(block->powHash), buf, 80, buf, 80, 1024, 1, 1);
-
+                break;
+                
+            default:
+                assert(0 && "Invalid algorithm");
```
Comment (MentalCollatz) [inline BRMerkleBlock.c:174]: Could this be used as a DoS attack vector?

Comment (SmartArray) [reply BRMerkleBlock.c:174]: Not if the asserts are disabled for productive builds -D NDEBUG

File: crypto/odocrypt.c (hunk 3 above)
Comment (SmartArray) [reply crypto/odocrypt.c]: Then we should remove 4 and 5 too :)

File: crypto/odocrypt.h (hunk 2 above)
Comment (SmartArray) [reply crypto/odocrypt.h]: Will do that in some hours 👍

File: crypto/odocrypt.h (hunk 1 above)
Comment (SmartArray) [reply crypto/odocrypt.h:63]: I know :3

Comment (SmartArray) [reply crypto/odocrypt.h:63]: Those are not exported

**Pull request merged**